src/audio.c

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (C) 2020  Joshua Minter & Patryk Seregiet
+ *
+ * This file is part of ReplaySorcery.
+ *
+ * ReplaySorcery is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * ReplaySorcery is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with ReplaySorcery.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "audio.h"
+#include "util/log.h"
+#include "util/memory.h"
+
+static void rsAudioReadSamples(RSAudio *audio);
+static void rsAudioPrepareEncoder(RSAudio *audio, const RSConfig* config);
+static void rsAudioEncoderPrepareFrame(RSAudioEncoder *audioenc);
+
+extern sig_atomic_t mainRunning;
+extern sig_atomic_t wantToSave;
+
+int rsAudioCreate(RSAudio *audio, RSConfig *config) {
+   uint8_t channels = config->audioChannels;
+   if (channels != 1 && channels != 2) {
+      rsLog("Only 1 or 2 channels are supported. Setting to 2");
+   }
+
+   int error;
+   const pa_sample_spec ss = {
+       .format = PA_SAMPLE_S16LE, 
+       .rate = config->audioSamplerate, 
+       .channels = channels
+   };
+   audio->paApi = pa_simple_new(NULL, "Replay Sorcery", PA_STREAM_RECORD, NULL,
+                                 "record", &ss, NULL, NULL, &error);
+
+   if (!audio->paApi) {
+      rsError("PulseAudio: pa_simple_new() failed: %s\n", pa_strerror(error));
+      return 0;
+   }
+   int size1s = pa_bytes_per_second(&ss);
+   int sizePerFrame = size1s / config->framerate;
+   int sizeTotal = size1s * config->duration;
```

Comment:
Would it be that bad of an idea to use `circle.c`? Yes its designed for dynamically sized data and this has a static size but it keeps code clean and reuses tested code

Reply:
I intended to do that at first, but circle.c doesn't do what I need. circle.c keeps a buffer of pointers to other buffers and you circle around those pointers. Make sense when you have a circle of individual frames that have different size (after jpeg compression). If I was to use  circle.c I would have to make new allocation for each frame of audio. Seems wasteful considering it's a small amount of data and the frames have constant size. One big buffer that I circle around is more efficient. Using the circle object would also make sync harder later on. If the video and audio gets out of sync I might to have to drop few samples and start encoding from the middle of a "frame". Since all frames are next to each other in one buffer it's not a problem, it's just a stream of audio data. Having it in separate buffers would introduce extra codes that copies some samples from this frame, others from the other.

Reply:
Maybe make another utility, called something like `static-circle.c` that handles this, along with handling the case where the next chunk of audio data gets split in two (half at the end half at the start) and has its own `Extract` function that realigns the data. Separation of concerns and all

Reply:
alright